### PR TITLE
Allow a proc to be used in addition to a static value for cookies_same_site_protection

### PR DESCRIPTION
**Pull request by JackMc**
### Summary

The cookies_same_site_protection configuration option is great, but our team has had trouble adopting it because it does not allow us to change which `SameSite` property value (if any) to use for a particular request. This PR adds the ability to specify a `Proc` that will determine which SameSite property value to use at run time.

### CODE REVIEW

**Review by rafaelfranca — COMMENTED** [review]
LGTM. Can you also update the documentation to show this can be done?

File: actionpack/lib/action_dispatch/middleware/cookies.rb [hunk 1]
```diff
@@ -437,7 +437,14 @@ def handle_options(options)
           end
 
           options[:path]      ||= "/"
-          options[:same_site] ||= request.cookies_same_site_protection
+
+          same_site_protection = if request.cookies_same_site_protection.is_a?(Proc)
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/middleware/cookies.rb]:
```suggestion
          same_site_protection = if request.cookies_same_site_protection.respond_to?(:call)
```

Comment (kaspth) [reply actionpack/lib/action_dispatch/middleware/cookies.rb]: Wonder if we want to auto convert a passed non-proc value into a proc when assigning the config?

Comment (rafaelfranca) [reply actionpack/lib/action_dispatch/middleware/cookies.rb]: We can do this.

File: railties/lib/rails/application.rb [hunk 2]
```diff
@@ -265,7 +271,7 @@ def env_config
           "action_dispatch.cookies_serializer" => config.action_dispatch.cookies_serializer,
           "action_dispatch.cookies_digest" => config.action_dispatch.cookies_digest,
           "action_dispatch.cookies_rotations" => config.action_dispatch.cookies_rotations,
-          "action_dispatch.cookies_same_site_protection" => config.action_dispatch.cookies_same_site_protection,
+          "action_dispatch.cookies_same_site_protection" => same_site_protection,
```
Comment (kaspth) [inline railties/lib/rails/application.rb]: Think we should extract a private helper method and pass in the config here.

```ruby
# … => coerce_same_site_protection(config.action_dispatch.cookies_same_site_protection)

private
  def coerce_same_site_protection(protection)
    protection.respond_to?(:call) ? protection : proc { protection }
  end
```

**Comment by rafaelfranca** [comment]
Merged in #40103

**Pull request closed without merging**